swift/ql/lib/codeql/swift/security/UncontrolledFormatStringQuery.qll

```diff
@@ -0,0 +1,23 @@
+/**
+ * Provides a taint-tracking configuration for reasoning about uncontrolled
+ * format string vulnerabilities.
+ */
+
+import swift
+import codeql.swift.StringFormat
+import codeql.swift.dataflow.DataFlow
+import codeql.swift.dataflow.TaintTracking
+import codeql.swift.dataflow.FlowSources
+
+/**
+ * A taint configuration for tainted data that reaches a format string.
+ */
+class TaintedFormatConfiguration extends TaintTracking::Configuration {
+  TaintedFormatConfiguration() { this = "TaintedFormatConfiguration" }
+
+  override predicate isSource(DataFlow::Node node) { node instanceof FlowSource }
```

Comment:
I think it's great that you're following the standard format for taint-flow queries like we started doing in https://github.com/github/codeql/pull/11086. However, it's still not _quite_ the same format as that PR. In particular, that PR added classes like:
```ql
/** A data flow sink for XML external entities (XXE) vulnerabilities. */
abstract class XxeSink extends DataFlow::Node { }

/** A sanitizer for XML external entities (XXE) vulnerabilities. */
abstract class XxeSanitizer extends DataFlow::Node { }

/**
 * A unit class for adding additional taint steps.
 *
 * Extend this class to add additional taint steps that should apply to paths related to
 * XML external entities (XXE) vulnerabilities.
 */
class XxeAdditionalTaintStep extends Unit {
  abstract predicate step(DataFlow::Node n1, DataFlow::Node n2);
}
```
so that we can tune both sources, sinks and additional taint steps without touching the configuration.

I think this PR should aim to do the same. Now that we're putting the configurations in the `.qll` file in the security directory we might as well follow the exact same format as the other languages that do the same.

Reply:
I'm a little unclear on what the sink pattern adds?  Is it just intended as a point for extension?  Should I create an abstract sanitizer even though I don't intend to create any concrete implementations at this time?

Reply:
Exactly. This is all to make it easier for people to extend the query's functionality. I do agree that I cannot really imagine another kind of sink for this particular query, but I think there's value in consistency.

Reply:
In the XXE and path injection queries by @atorralba, we actually have the codeql split across three files.  For a query `Abc` we have:
- `Abc.ql`, containing just the metadata and from-where-select
- `AbcQuery.qll`, containing just the dataflow configuration
- `Abc.qll`, containing sinks + sanitizers + additional flow steps

I'm a bit concerned about this proliferation of files being difficult to navigate.  I'd like to discuss two alternatives (even if only to educate myself about why we don't want to do them):
1. put the dataflow configuration, sinks, sanitizers and additional flow steps all together in one library file, `AbcQuery.qll`.  This cuts us back to two files containing codeql per query, and as far as I can tell a user can still customize the query and re-use components when they need to.
2. stick to three files, but rename the third one `AbcModels.qll`.  I want it to be clearer what that file's about.

Note that in the query in this PR I currently do have a third codeql file, `StringFormat.qll`.  But that is *not* sinks and sanitizers, it's a re-usable library for reasoning about format strings (similar to but currently greatly simpler than the CPP `FormattingFunction.qll`).  In another query those things could be sources, or sanitizers, or something else.

Reply:
> put the dataflow configuration, sinks, sanitizers and additional flow steps all together in one library file, `AbcQuery.qll`. This cuts us back to two files containing codeql per query, and as far as I can tell a user can still customize the query and re-use components when they need to.

If we put them in one file then it's not possible for a user to import the sources, sinks and sanitizers from one query into another without _also_ importing the dataflow configuration. And importing an additional dataflow configuration will incur a serious performance overhead.

By having the dataflow configuration in its own file a user can write a query that combines the sources, sinks and sanitizers from two completely independent queries in one query like:
```ql
import codeql.swift.security.UnsafeFoo as Foo
import codeql.swift.security.UnsafeBar as Bar
import ...

class MySource extends DataFlow::Node {
  MySource() {
    this instanceof Foo::SourceNode or this instanceof Bar::SourceNode
  }
}

class MyConf extends DataFlow::Configuration { ... }
```
If importing `UnsafeFoo` also import an additional dataflow configuration, then this pattern will not be possible (like it is in the other languages that apply this pattern).

Reply:
> 2. stick to three files, but rename the third one AbcModels.qll. I want it to be clearer what that file's about.

I can see where you're coming from with this. Two counterpoints:
1. I personally value cross-language consistency much more than the specific name of the file. Being able to tell customers that query layout for Swift queries is the same as Python's and C#'s (and the other languages that apply this pattern) is a really strong argument!
2. I'm not sure I like the word `Models` here, though. Models usually refer to specific models for functions (for example, a summary of its dataflow behavior), and I think it'll be confusing to start using the word "Model" to mean sources, sinks, and sanitizers as well.

Reply:
`AbcExtensions.qll` is better than just `Abc.qll`.  I can live with it.

I need to update this PR, and write up a ticket to make all of our queries consistent.

Reply:
I would also like to reach cross-language consistency at some point, like @MathiasVP says. Is this something other language teams can be convinced of?

Reply:
I've created https://github.com/github/codeql-c-team/issues/1450 for modernizing our existing queries, though limited to the 'injection' style ones that I think will benefit the most from this.

I've created https://github.com/github/codeql-c-team/issues/1451 for the sensitive expressions (`cleartext-*`) queries, though I consider it lower priority.

I'm happy to participate in any cross-language discussion about this, but I'm not very motivated to drive such a conversation right now.

Just need to update this PR now...

Reply:
I've just updated this PR to the new way (latest two commits).  Please do nit-pick as any mistakes are likely to be copied in future queries.